File: src/main/java/com/heybys/oddments/fooddelivery/domain/generic/Money.java
package com.heybys.oddments.fooddelivery.domain.generic;

import java.math.BigDecimal;
import java.util.Collection;
import java.util.function.Function;

import com.heybys.oddments.base.domain.ValueObject;

import lombok.Getter;

@Getter
public class Money extends ValueObject<Money> {
    public static final Money ZERO = Money.wons(0);

    private final BigDecimal amount;

    public Money(BigDecimal amount) {
        this.amount = amount;
    }

    public static Money wons(long amount) {
        return new Money(BigDecimal.valueOf(amount));
    }

    public static Money wons(double amount) {
        return new Money(BigDecimal.valueOf(amount));
    }

    public static <T> Money sum(Collection<T> bags, Function<T, Money> monetary) {
        return bags.stream().map(monetary).reduce(Money.ZERO, Money::plus);
    }

    @Override
    public boolean equals(Object other) {
        return super.equals(other);
    }

    @Override
    public int hashCode() {
        return super.hashCode();
    }

    @Override
    protected Object[] getEqualityFields() {
        return new Object[] {amount.doubleValue()};
    }

    public Money plus(Money amount) {
        return new Money(this.amount.add(amount.amount));
    }

    public Money minus(Money amount) {
        return new Money(this.amount.subtract(amount.amount));
    }

    public Money times(double percent) {
        return new Money(this.amount.multiply(BigDecimal.valueOf(percent)));
    }

    public Money divide(double divisor) {
        return new Money(amount.divide(BigDecimal.valueOf(divisor)));
    }

    public boolean isLessThan(Money other) {
        return amount.compareTo(other.amount) < 0;
    }

    public boolean isGreaterThanOrEqual(Money other) {
        return amount.compareTo(other.amount) >= 0;
    }

    public Long longValue() {
        return amount.longValue();
    }

    public Double doubleValue() {
        return amount.doubleValue();
    }

    public String toString() {
        return amount.toString() + "원";
    }
}

File: src/main/kotlin/com/heybys/oddments/fooddelivery/domain/generic/Money.kt
package com.heybys.oddments.fooddelivery.domain.generic

import com.heybys.oddments.base.domain.ValueObject
import java.math.BigDecimal
import java.util.function.Function

class Money(val amount: BigDecimal) : ValueObject<Money>() {

    companion object {
        @JvmField val ZERO = Money.wons(0)

        @JvmStatic fun wons(amount: Long): Money = Money(BigDecimal.valueOf(amount))

        @JvmStatic fun wons(amount: Double): Money = Money(BigDecimal.valueOf(amount))

        @JvmStatic
        fun <T> sum(bags: Collection<T>, monetary: Function<T, Money>): Money =
            bags.map { monetary.apply(it) }.fold(ZERO) { acc, money -> acc.plus(money) }
    }

    override fun equals(other: Any?): Boolean = super.equals(other)

    override fun hashCode(): Int = super.hashCode()

    override val equalityFields: Array<Any?>
        get() = arrayOf(amount.toDouble())

    fun plus(amount: Money): Money = Money(this.amount.add(amount.amount))

    fun minus(amount: Money): Money = Money(this.amount.subtract(amount.amount))

    fun times(percent: Double): Money = Money(this.amount.multiply(BigDecimal.valueOf(percent)))

    fun divide(divisor: Double): Money = Money(amount.divide(BigDecimal.valueOf(divisor)))

    fun isLessThan(other: Money): Boolean = amount.compareTo(other.amount) < 0

    fun isGreaterThanOrEqual(other: Money): Boolean = amount.compareTo(other.amount) >= 0

    fun longValue(): Long = amount.toLong()

    fun doubleValue(): Double = amount.toDouble()

    override fun toString(): String = "${amount}원"
}
